File: typescript/src/hooks/mapHookState.ts
import { HookState } from './types';
import { HookStateExitFee } from './exitFeeHook';
import { HookStateStableSurge } from './stableSurgeHook';

type HookData = {
    address: string;
    type: string;
    enableHookAdjustedAmounts: boolean;
    shouldCallAfterSwap: boolean;
    shouldCallBeforeSwap: boolean;
    shouldCallAfterInitialize: boolean;
    shouldCallBeforeInitialize: boolean;
    shouldCallAfterAddLiquidity: boolean;
    shouldCallBeforeAddLiquidity: boolean;
    shouldCallAfterRemoveLiquidity: boolean;
    shouldCallBeforeRemoveLiquidity: boolean;
    shouldCallComputeDynamicSwapFee: boolean;
    dynamicData?: Record<string, string>;
};

/**
 * Maps hook data from JSON test files into typed HookState objects
 * that can be used with hook implementations
 */
export function mapHookState(
    hookData: HookData,
    poolData: { tokens: string[]; amp?: bigint },
): HookState {
    switch (hookData.type) {
        case 'EXIT_FEE':
            return mapExitFeeHookState(hookData, poolData.tokens);
        case 'STABLE_SURGE':
            return mapStableSurgeHookState(hookData, poolData);
        default:
            throw new Error(`Unsupported hook type: ${hookData.type}`);
    }
}

function mapExitFeeHookState(
    hookData: HookData,
    tokens: string[],
): HookStateExitFee {
    if (!hookData.dynamicData?.removeLiquidityHookFeePercentage) {
        throw new Error(
            'EXIT_FEE hook requires removeLiquidityHookFeePercentage in dynamicData',
        );
    }

    const {
        shouldCallComputeDynamicSwapFee,
        shouldCallBeforeSwap,
        shouldCallAfterSwap,
        shouldCallBeforeAddLiquidity,
        shouldCallAfterAddLiquidity,
        shouldCallBeforeRemoveLiquidity,
        shouldCallAfterRemoveLiquidity,
        enableHookAdjustedAmounts,
    } = hookData;

    return {
        hookType: 'ExitFee',
        // Configuration flags
        shouldCallComputeDynamicSwapFee,
        shouldCallBeforeSwap,
        shouldCallAfterSwap,
        shouldCallBeforeAddLiquidity,
        shouldCallAfterAddLiquidity,
        shouldCallBeforeRemoveLiquidity,
        shouldCallAfterRemoveLiquidity,
        enableHookAdjustedAmounts,
        // Hook-specific data
        tokens,
        removeLiquidityHookFeePercentage: BigInt(
            hookData.dynamicData.removeLiquidityHookFeePercentage,
        ),
    };
}

function mapStableSurgeHookState(
    hookData: HookData,
    poolData: { tokens: string[]; amp?: bigint },
): HookStateStableSurge {
    if (!hookData.dynamicData?.surgeThresholdPercentage) {
        throw new Error(
            'STABLE_SURGE hook requires surgeThresholdPercentage in dynamicData',
        );
    }
    if (!hookData.dynamicData?.maxSurgeFeePercentage) {
        throw new Error(
            'STABLE_SURGE hook requires maxSurgeFeePercentage in dynamicData',
        );
    }
    if (!poolData.amp) {
        throw new Error('STABLE_SURGE hook requires amp from pool data');
    }

    const {
        shouldCallComputeDynamicSwapFee,
        shouldCallBeforeSwap,
        shouldCallAfterSwap,
        shouldCallBeforeAddLiquidity,
        shouldCallAfterAddLiquidity,
        shouldCallBeforeRemoveLiquidity,
        shouldCallAfterRemoveLiquidity,
        enableHookAdjustedAmounts,
    } = hookData;

    return {
        hookType: 'StableSurge',
        // Configuration flags
        shouldCallComputeDynamicSwapFee,
        shouldCallBeforeSwap,
        shouldCallAfterSwap,
        shouldCallBeforeAddLiquidity,
        shouldCallAfterAddLiquidity,
        shouldCallBeforeRemoveLiquidity,
        shouldCallAfterRemoveLiquidity,
        enableHookAdjustedAmounts,
        // Hook-specific data
        amp: poolData.amp,
        surgeThresholdPercentage: BigInt(
            hookData.dynamicData.surgeThresholdPercentage,
        ),
        maxSurgeFeePercentage: BigInt(
            hookData.dynamicData.maxSurgeFeePercentage,
        ),
    };
}

File: python/src/hooks/map_hook_state.py
from src.hooks.exit_fee.types import ExitFeeHookState, map_exit_fee_hook_state
from src.hooks.stable_surge.types import StableSurgeHookState, map_stable_surge_hook_state
from src.hooks.types import HookState


def map_hook_state(hook_data: dict, pool_data: dict) -> HookState:
    """
    Maps hook data from JSON to typed HookState objects.

    This function serves as a central dispatcher that converts raw hook data
    from test JSON files into strongly-typed HookState objects based on the
    hook type.

    Args:
        hook_data: Raw hook dict from JSON with fields:
            - type: Hook type identifier (e.g., "STABLE_SURGE", "EXIT_FEE")
            - dynamicData: Hook-specific parameters (varies by type)
            - address: Hook contract address
            - Configuration flags (shouldCallAfterSwap, etc.)
        pool_data: Pool dict with fields needed by hooks:
            - tokens: List of token addresses
            - amp: Amplification parameter (for stable pools with STABLE_SURGE)

    Returns:
        Typed HookState object (StableSurgeHookState or ExitFeeHookState)

    Raises:
        ValueError: If hook type is unsupported or required fields are missing
    """
    hook_type = hook_data.get("type")

    if hook_type == "STABLE_SURGE":
        return _map_stable_surge_hook(hook_data, pool_data)
    elif hook_type == "EXIT_FEE":
        return _map_exit_fee_hook(hook_data, pool_data)
    else:
        raise ValueError(f"Unsupported hook type: {hook_type}")


def _map_stable_surge_hook(hook_data: dict, pool_data: dict) -> StableSurgeHookState:
    """Maps STABLE_SURGE hook data to StableSurgeHookState."""
    if "dynamicData" not in hook_data:
        raise ValueError("STABLE_SURGE hook requires dynamicData")

    dynamic_data = hook_data["dynamicData"]

    if "surgeThresholdPercentage" not in dynamic_data:
        raise ValueError(
            "STABLE_SURGE hook requires surgeThresholdPercentage in dynamicData"
        )
    if "maxSurgeFeePercentage" not in dynamic_data:
        raise ValueError(
            "STABLE_SURGE hook requires maxSurgeFeePercentage in dynamicData"
        )
    if "amp" not in pool_data:
        raise ValueError("STABLE_SURGE hook requires amp from pool data")

    return map_stable_surge_hook_state(
        {
            "surgeThresholdPercentage": dynamic_data["surgeThresholdPercentage"],
            "maxSurgeFeePercentage": dynamic_data["maxSurgeFeePercentage"],
            "amp": pool_data["amp"],
        }
    )


def _map_exit_fee_hook(hook_data: dict, pool_data: dict) -> ExitFeeHookState:
    """Maps EXIT_FEE hook data to ExitFeeHookState."""
    if "tokens" not in pool_data:
        raise ValueError("EXIT_FEE hook requires tokens from pool data")

    return map_exit_fee_hook_state(hook_data, pool_data["tokens"])
